sstables/sstable_set.cc

```diff
@@ -727,6 +727,67 @@ sstable_set_impl::create_single_key_sstable_reader(
     return make_combined_reader(schema, std::move(permit), std::move(readers), fwd, fwd_mr);
 }
 
+flat_mutation_reader
+time_series_sstable_set::create_single_key_sstable_reader(
+        column_family* cf,
+        schema_ptr schema,
+        reader_permit permit,
+        utils::estimated_histogram& sstable_histogram,
+        const dht::ring_position& pos,
+        const query::partition_slice& slice,
+        const io_priority_class& pc,
+        tracing::trace_state_ptr trace_state,
+        streamed_mutation::forwarding fwd_sm,
+        mutation_reader::forwarding fwd_mr) const {
+    // First check if the optimized algorithm for TWCS single partition queries can be applied.
+    // Multiple conditions must be satisfied:
+    // 1. The sstables must be sufficiently modern so they contain the min/max column metadata.
+    // 2. The schema cannot have static columns, since we're going to be opening new readers
+    //    into new sstables in the middle of the partition query. TWCS sstables will usually pass
+    //    this condition.
+    // 3. The sstables cannot have partition tombstones for the same reason as above.
+    //    TWCS sstables will usually pass this condition.
+    using sst_entry = std::pair<position_in_partition, shared_sstable>;
+    if (schema->has_static_columns()
+            || std::any_of(_sstables->begin(), _sstables->end(),
+                [] (const sst_entry& e) {
+                    return e.second->get_version() < sstable_version_types::md
+                        || e.second->may_have_partition_tombstones();
+    })) {
+        // Some of the conditions were not satisfied so we use the standard query path.
+        return sstable_set_impl::create_single_key_sstable_reader(
+                cf, std::move(schema), std::move(permit), sstable_histogram,
+                pos, slice, pc, std::move(trace_state), fwd_sm, fwd_mr);
+    }
+
+    auto pk_filter = make_pk_filter(pos, *schema);
+    auto it = std::find_if(_sstables->begin(), _sstables->end(), [&] (const sst_entry& e) { return pk_filter(*e.second); });
+    if (it == _sstables->end()) {
+        // No sstables contain data for the queried partition.
+        return make_empty_flat_reader(std::move(schema), std::move(permit));
+    }
+
+    auto ck_filter = [ranges = slice.get_all_ranges()] (const sstable& sst) { return sst.may_contain_rows(ranges); };
+    it = std::find_if(it, _sstables->end(), [&] (const sst_entry& e) { return ck_filter(*e.second); });
+    if (it == _sstables->end()) {
```

Comment:
You could have filtered here, instead of letting the reader_queue do the filtering again, saving a pass.

Reply:
(both the pk and ck filtering could have been done here, instead of just a check)

Reply:
I would need to copy the `shared_sstable`s, to a new `multimap`, which I wanted to avoid.
But you're right, there's an unnecessary double pass with the filter...
There's another way to fix it: pass the iterator to `min_position_reader_queue` constructor